### ISSUE TITLE
Transaction based database cleanup for tests

### ISSUE THREAD

**Issue by pex** [vitaly-t/pg-promise#538]
Transaction strategy is a common method to have a clean database for each test run. Which means a transaction is created before and rolled back after each test run.

I'd like to explicitly rollback or commit transactions to achieve this. Maybe we could add an option parameter to switch between auto and manual transactions?

### Expected behavior

```js
// Pseudo example with Jest

jest.mock('./db.js')

const db = require('./db.js') // Mocked adapter used by app
const originalDb = require.requireActual('./db.js')

beforeEach(() => {
  originalDb().tx(tx => {
    db.mockReturnValue(tx) // All queries now run on tx
  })
})

afterEach(() => {
  db().rollback() // tx has to be rollbacked
})
```

### Actual behavior

`pg-promise` doesn't support explicit rollbacks

**Comment by vitaly-t** [comment]
A transaction in this library manages its integrity by fully encapsulating the connection inside it, as well as result of the operation. What you are asking, is to compromise both of those things, by exposing the connection outside of the transaction, and its ability to handle errors correctly.

For any application it should be strictly a no-no.

But if you want to hack your tests that way, you can manage the connection manually all the way, but be careful not to leak the connections, and to handle errors properly...

```js
let cn;

try {
    cn = await db.connect(); // manually created connection;   

    await cn.query('BEGIN');

    /* and so on */

    await cn.query('COMMIT'); // or `ROLLBACK`

} catch(error) {

// handle errors

} finally() {
    cn.done(); // must release the connection
}
```

You can do all that, but it isn't a good idea.

**Comment by vitaly-t** [comment]
You are looking at testing database integrity the wrong way. The right way is to test for data that was either rolled back or committed. Then you do not need to reinvent those dangerous things here.

**Comment by clintonmedbery** [comment]
@vitaly-t I am not sure you are understanding the request, or perhaps I don't understand your answer. When I am writing tests I will insert a user for example and maybe other rows to test against. When I am done I want to remove them so I can test with a clean db. It would be nice to be able to roll these back. I don't need to test rollback functionality. Is there a solution for this?

**Comment by vitaly-t** [comment]
@clintonmedbery You should use a transaction per test then, because transaction logic doesn't map well over multiple tests, being connection-dependent.

**Comment by bheemreddy181** [comment]
I assume everyone is looking for a similar strategy to test the functional tests with data being inserted to the DB and cleaned, if by any chance two tests use the same data that would cause a problem in general. Database cleaning strategy at the transactional level would be really great something how they do in ruby with database cleaner gem https://github.com/DatabaseCleaner/database_cleaner#how-to-use

**Comment by bheemreddy181** [comment]
Yes, we want to use transaction per test and roll that back after each test. So DB is always clean

**Comment by noinkling** [comment]
> You should use a transaction per test then, because transaction logic doesn't map well over multiple tests

Even that's pretty awkward with the `.tx` API, for example in Jest:

```js
test('some test', async () => {
  
  // Not actually necessary in this case, but it others it might be
  expect.hasAssertions();

  try {
    await db.tx(async (tx) => {
      // Contrived example
      await tx.none(`INSERT INTO users (foo, bar) VALUES ('hello', 123)`);

      const count = await tx.one(`SELECT count(*) FROM users`);
      expect(count).toBe('1');
  
      throw 'INTENTIONAL ROLLBACK';
    });
  } catch (err) {
    // Have to catch the intentional rollback so it doesn't cause the test to
    // fail, but failing assertions and other errors need to be propagated
    if (err !== 'INTENTIONAL ROLLBACK') throw err;
  }
});
```

**Comment by noinkling** [comment]
After some experimentation, I discovered that the `db.connect()` approach is problematic when nested tasks/transactions are involved because they don't seem to reuse the connection/session (nor are they aware of the outer transaction), presumably because the connection object doesn't have the `ctx` that other protocol objects have. That basically limits you to using `.tx()` if you want a robust solution.

That's ok if you're doing per-test transactions, you can just make a wrapper along the lines of what I posted above:

```js
const INTENTIONAL_ROLLBACK = Symbol('ROLLBACK')

async function withRolledBackTx(callback, txOptions, base = db) {
  try {
    await base.tx(txOptions, async (tx) => {
      await callback(tx)
      throw INTENTIONAL_ROLLBACK
    })
  } catch (error) {
    if (error !== INTENTIONAL_ROLLBACK) throw error
  }
}

// In tests:
test('whatever', () => {
  return withRolledBackTx(async (tx) => {
    // whatever
  })
})
```
There are ways to make this more ergonomic so you could just write `test('whatever', rollback((tx) => { ... }))` or even `test.rollback('whatever', (tx) => { ... })`

If you want to share a single transaction across tests, it's a lot trickier, but I came up with a way for the `tx` to "escape" the callback, and to be explicitly committed or rolled back:

```js
function useTx(txOptions, base = db) {
  return new Promise((resolve, reject) => {
    const txCompletePromise = base.tx(txOptions, function (tx) {
      return new Promise((commit, rollback) => {
        let proxy
        try {
          const revocable = Proxy.revocable(tx, {
            get(target, property, receiver) {
              if (property === 'rollback') {
                return async () => {
                  rollback()
                  await txCompletePromise
                  revocable.revoke()
                }
              }
              if (property === 'commit') {
                return async () => {
                  commit()
                  await txCompletePromise
                  revocable.revoke()
                }
              }
              return target[property]
            }
          })
          proxy = revocable.proxy
        } catch (error) {
          reject(error)
          throw error
        }
        resolve(proxy)
      })
    }).catch(() => {})
  })
}


// In tests:
let tx

beforeAll(async () => {
  tx = await useTx()
})

afterAll(async () => {
  await tx.rollback()
})

test('whatever', async () => {
  // use `tx` like you would normally
})
```
Note that because the `tx` object is frozen, you can't monkey-patch it directly with the `rollback` and `commit` methods. In this case I'm using a (revocable) `Proxy` as a workaround, but there are other, probably cleaner, approaches you could use.

You could also use `useTx()` in `beforeEach()` as an alternative to the wrapper approach, just be aware that using a shared variable like this example would be problematic if your tests are set up to run concurrently (in Jest they run sequentially by default, it's only other files that run concurrently). Some frameworks might expose a per-test context inside `beforeEach()` (or equivalent) to get around this limitation, but I'd recommend just sticking with the wrapper approach.

Hopefully this is helpful to somebody.
